app/assets/javascripts/discourse/app/static/prosemirror/lib/text-manipulation.js

```diff
@@ -0,0 +1,492 @@
+    this.view.dispatch(
+      this.view.state.tr
+        .replaceSelectionWith(this.schema.nodes.emoji.create({ code }))
+        .insertText(" ")
+    );
+  }
+
+  @bind
+  paste() {
+    // Intentionally no-op
+    // Pasting markdown is being handled by the markdown-paste extension
+    // Pasting a url on top of a text is being handled by the link extension
+  }
+
+  selectText(from, length, opts) {
+    const tr = this.view.state.tr.setSelection(
+      new TextSelection(
+        this.view.state.doc.resolve(from),
+        this.view.state.doc.resolve(from + length)
+      )
+    );
+
+    if (opts.scroll) {
+      tr.scrollIntoView();
+    }
+
+    this.view.dispatch(tr);
+  }
+
+  @bind
+  inCodeBlock() {
+    return this.autocompleteHandler.inCodeBlock();
+  }
+
+  indentSelection(direction) {
+    const { selection } = this.view.state;
+
+    const isInsideListItem =
+      selection.$head.depth > 0 &&
+      selection.$head.node(-1).type === this.schema.nodes.list_item;
+
+    if (isInsideListItem) {
+      const command =
+        direction === "right"
+          ? sinkListItem(this.schema.nodes.list_item)
+          : liftListItem(this.schema.nodes.list_item);
+      command(this.view.state, this.view.dispatch);
+      return true;
+    }
+  }
+
+  insertText(text) {
+    const doc = this.convertFromMarkdown(text);
+
+    this.view.dispatch(
+      this.view.state.tr
+        .replaceSelectionWith(doc.content.firstChild)
+        .scrollIntoView()
+    );
+
+    this.focus();
+  }
+
+  replaceText(oldValue, newValue, opts = {}) {
+    // Replacing Markdown text is not reliable and should eventually be deprecated
+
+    const markdown = this.convertToMarkdown(this.view.state.doc);
+
+    const regex = opts.regex || new RegExp(oldValue, "g");
+    const index = opts.index || 0;
+    let matchCount = 0;
+
+    const newMarkdown = markdown.replace(regex, (match) => {
+      if (matchCount++ === index) {
+        return newValue;
+      }
+      return match;
+    });
+
+    if (markdown === newMarkdown) {
+      return;
+    }
+
+    const newDoc = this.convertFromMarkdown(newMarkdown);
+    if (!newDoc) {
+      return;
+    }
+
+    const diff = newValue.length - oldValue.length;
+    const startOffset = this.view.state.selection.from + diff;
+    const endOffset = this.view.state.selection.to + diff;
+
+    const tr = this.view.state.tr.replaceWith(
+      0,
+      this.view.state.doc.content.size,
+      newDoc.content
+    );
+
+    if (
+      !opts.skipNewSelection &&
+      (opts.forceFocus || this.view.dom === document.activeElement)
+    ) {
+      const adjustedStart = Math.min(startOffset, tr.doc.content.size);
+      const adjustedEnd = Math.min(endOffset, tr.doc.content.size);
+
+      tr.setSelection(TextSelection.create(tr.doc, adjustedStart, adjustedEnd));
+    }
+
+    this.view.dispatch(tr);
+  }
+
+  toggleDirection() {
+    this.view.dom.dir = this.view.dom.dir === "rtl" ? "ltr" : "rtl";
+  }
```

Comment:
With WYSIWYG, I think we might actually be able to drop this button. Users who want to mix LTR and RTL languages should be able to do it using the native unicode stuff. (that's not supported in textarea, which is why we had to add this dir-flipping button)

Fine to leave it in this PR, but we should ask one of our RTL-language-speaking collegues for advice on whether it can be removed in future 🤞